File: plugins/opencortex-memory/hooks/hook_runner.py
#!/usr/bin/env python3
"""Cross-platform hook dispatcher for OpenCortex Claude Code plugin.

Detects the current platform and runs the appropriate hook script:
- macOS/Linux: bash <hook>.sh
- Windows: powershell -NoProfile -ExecutionPolicy Bypass -File <hook>.ps1

Usage:
    python3 hook_runner.py <hook-name>
    e.g. python3 hook_runner.py session-start
"""

import os
import platform
import subprocess
import sys


def main():
    if len(sys.argv) < 2:
        print('{"error": "hook name required"}', file=sys.stderr)
        sys.exit(1)

    hook_name = sys.argv[1]
    hooks_dir = os.path.dirname(os.path.abspath(__file__))

    # Read stdin (hook input JSON) before spawning subprocess
    stdin_data = b""
    if not sys.stdin.isatty():
        try:
            stdin_data = sys.stdin.buffer.read()
        except Exception:
            pass

    if platform.system() == "Windows":
        script = os.path.join(hooks_dir, f"{hook_name}.ps1")
        if not os.path.isfile(script):
            print(f'{{"error": "script not found: {hook_name}.ps1"}}')
            sys.exit(1)
        cmd = [
            "powershell", "-NoProfile", "-ExecutionPolicy", "Bypass",
            "-File", script,
        ]
    else:
        script = os.path.join(hooks_dir, f"{hook_name}.sh")
        if not os.path.isfile(script):
            print(f'{{"error": "script not found: {hook_name}.sh"}}')
            sys.exit(1)
        cmd = ["bash", script]

    # Propagate environment variables that Claude Code sets
    env = os.environ.copy()
    plugin_root = env.get("CLAUDE_PLUGIN_ROOT", os.path.dirname(hooks_dir))
    env.setdefault("CLAUDE_PLUGIN_ROOT", plugin_root)

    result = subprocess.run(
        cmd,
        input=stdin_data,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        env=env,
    )

    # Forward stdout (hook JSON output) and stderr
    if result.stdout:
        sys.stdout.buffer.write(result.stdout)
        sys.stdout.buffer.flush()
    if result.stderr:
        sys.stderr.buffer.write(result.stderr)
        sys.stderr.buffer.flush()

    sys.exit(result.returncode)


if __name__ == "__main__":
    main()

File: plugins/opencortex-memory/hooks/hook_runner.mjs
#!/usr/bin/env node
/**
 * Cross-platform hook dispatcher for OpenCortex Claude Code plugin.
 *
 * Detects platform and spawns the correct shell script:
 *   - macOS/Linux → bash <hook>.sh
 *   - Windows     → powershell -NoProfile -ExecutionPolicy Bypass -File <hook>.ps1
 *
 * Usage: node hook_runner.mjs <hook-name>
 *   e.g. node hook_runner.mjs session-start
 */

import { spawn } from "node:child_process";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { platform } from "node:os";

const __dirname = dirname(fileURLToPath(import.meta.url));
const hookName = process.argv[2];

if (!hookName) {
  process.stderr.write('{"error": "hook name required"}\n');
  process.exit(1);
}

// Collect stdin before spawning child
const chunks = [];
process.stdin.on("data", (chunk) => chunks.push(chunk));
process.stdin.on("end", () => {
  const stdinData = Buffer.concat(chunks);

  let cmd, args;
  if (platform() === "win32") {
    const script = join(__dirname, `${hookName}.ps1`);
    cmd = "powershell";
    args = ["-NoProfile", "-ExecutionPolicy", "Bypass", "-File", script];
  } else {
    const script = join(__dirname, `${hookName}.sh`);
    cmd = "bash";
    args = [script];
  }

  const child = spawn(cmd, args, {
    env: {
      ...process.env,
      CLAUDE_PLUGIN_ROOT:
        process.env.CLAUDE_PLUGIN_ROOT || join(__dirname, ".."),
    },
    stdio: ["pipe", "pipe", "pipe"],
  });

  child.stdin.write(stdinData);
  child.stdin.end();

  child.stdout.pipe(process.stdout);
  child.stderr.pipe(process.stderr);

  child.on("close", (code) => process.exit(code ?? 0));
});

// Handle empty stdin (no piped data)
if (process.stdin.isTTY) {
  process.stdin.emit("end");
}
